testing/cpm/verify_generated_pins.cmake

```
@@ -0,0 +1,46 @@
+#=============================================================================
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+
+function(verify_generated_pins target_name)
+  set(_rapids_options)
+  set(_rapids_one_value PIN_FILE)
+  set(_rapids_multi_value PROJECTS PROJECTS_NOT_EXIST)
+  cmake_parse_arguments(_RAPIDS "${_rapids_options}" "${_rapids_one_value}"
+                        "${_rapids_multi_value}" ${ARGN})
+
+  if(NOT DEFINED _RAPIDS_PROJECTS)
+    message(FATAL_ERROR "verify_generated_pins must be called with `PROJECTS` to verify")
+  endif()
+  if(NOT DEFINED _RAPIDS_PIN_FILE)
+    set(_RAPIDS_PIN_FILE "${CMAKE_CURRENT_BINARY_DIR}/rapids-cmake/pinned_versions.json")
+  endif()
+
+  # only check projects that were downloaded by CPM (ignore those already in the build environment)
+  foreach(proj IN LISTS _RAPIDS_PROJECTS)
+    if(${proj}_SOURCE_DIR)
+      list(APPEND projects-to-verify ${proj})
+    endif()
+  endforeach()
+
+  add_custom_target(${target_name} ALL
+    COMMAND ${CMAKE_COMMAND} -S="${CMAKE_CURRENT_FUNCTION_LIST_DIR}/verify_generated_pins/" -B"${CMAKE_BINARY_DIR}/${target_name}_verify_build"
+    "-Drapids-cmake-dir=${rapids-cmake-dir}"
+    "-Dpinned_versions_file=${_RAPIDS_PIN_FILE}"
+    "-Dprojects-to-verify=\"${projects-to-verify}\""
+    "-Dprojects-not-in-list=\"${_RAPIDS_PROJECTS_NOT_EXIST}\""
```

```suggestion
    "-Dprojects-to-verify=${projects-to-verify}"
    "-Dprojects-not-in-list=${_RAPIDS_PROJECTS_NOT_EXIST}"
```

---

This will break since the values are lists, and lists need to be quoted or the `;` will be captured by the shell instead.

```
bash-3.2$ pwd
/Users/rmaynard/Work/temp/custom_target_args/build
bash-3.2$ cmake -Dtest=A;B;C -S ../
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/Users/rmaynard/Work/temp/custom_target_args/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
bash: B: command not found
bash: C: command not found
```

---

Adding the `VERBATIM` option helps (frankly, this should really be the default anyway):

```suggestion
    "-Dprojects-to-verify=${projects-to-verify}"
    "-Dprojects-not-in-list=${_RAPIDS_PROJECTS_NOT_EXIST}"
    VERBATIM
```